### ISSUE TITLE
Items added to Globals come through in editor but not in build

### ISSUE THREAD

**Issue by amagee** [ReactUnity/core#29]
If I add some items to a `ReactUnityUGUI`'s `Globals` in a script (similarly to how I did in https://github.com/ReactUnity/core/issues/28):

```
public class MyComponent : MonoBehaviour {
    public void SetupStuff() {
        var component = GameObject.Find("SomeObject").GetComponent<ReactUnity.UGUI.ReactUnityUGUI>();

        component.Globals["someProp"] = "some value";
        component.Globals["someOtherProp"] = <some other value, it doesn't seem to matter>;
    }
}
```

and in my React I use `globalsWatcher.useContext()` to try to access them, everything works fine when I'm running the game from the Unity editor.

However when I create an actual build, the objects are not there on the Javascript side. There's no error that I can see, the objects are just absent.

Objects that I add to the Globals statically via the inspector in the Unity UI do work both in the editor and in the build, though.

**Comment by KurtGokhan** [comment]
I have tests specifically for this, and it seems to be working.

Let's diagnose the issue correctly. Changes to the Globals object may not be reflected on the React side until the next frame. So, if you assign globals like this, that value will not be available for at least 1 frame. Maybe you are experiencing this?

ReactUnity does things mostly the same for Editor and Build. The only difference that can cause this is, in the Build the websocket (and devserver) will be disabled by default. You can add `REACT_ALLOW_DEVSERVER` to the define symbols and see if that changes anything.

**Comment by KurtGokhan** [comment]
You can also try disabling `ReactUnityUGUI` component initially and enable it after Globals are assigned, just to see if that solves the problem.

Another thing you can try is, assigning the values to `component.Context.Globals` instead of `component.Globals`. This is what assigning to `component.Globals` does indirectly, but in case there is an error there, this should be the solution.

**Comment by KurtGokhan** [comment]
I forgot to mention one important thing which may be the reason for your other issue (and this one as well). When building with IL2CPP, Unity will strip out unused code to make the build size smaller and faster. And because ReactUnity uses reflection to reach code, Unity cannot detect that code and will strip it away as well. You can read more about code stripping here: https://docs.unity3d.com/Manual/ManagedCodeStripping.html

The easy way to solve this is adding `Preserve` attribute to methods you use in code:

`
[UnityEngine.Scripting.Preserve]
public void SetupStuff() {
}
`

Another way is to add `[assembly: UnityEngine.Scripting.Preserve]` to anywhere in your code to prevent stripping in all of that assembly.

Unity can also strip unused engine code, but that code may be used by ReactUnity. The way to prevent that is by using a `link.xml` file. This is an example `link.xml` file I used because Unity was stripping away code of Video component:

```
<linker>
  <assembly fullname="UnityEngine.VideoModule">
    <type fullname="UnityEngine.Video.VideoPlayer" preserve="all"/>
  </assembly>

  <assembly fullname="Assembly-CSharp"/>
  
  <assembly fullname="ReactUnity"/>
</linker>
```

**Comment by amagee** [comment]
Hey, thanks (again) for your quick and thorough response, it is appreciated.

Here are my results with your suggestions:

> Let's diagnose the issue correctly. Changes to the Globals object may not be reflected on the React side until the next frame. So, if you assign globals like this, that value will not be available for at least 1 frame. Maybe you are experiencing this?

No, this isn't the issue. I am waiting several seconds and the globals are still not available after that time.

> You can add REACT_ALLOW_DEVSERVER to the define symbols and see if that changes anything.

Yes, this fixes the issue.

> You can also try disabling ReactUnityUGUI component initially and enable it after Globals are assigned, just to see if that solves the problem.

This also fixes the issue (independently from the above - so REACT_ALLOW_DEVSERVER is not required if I do this).
Interestingly, if I do this, I can make additional updates after enabling the component and they also work.

> Another thing you can try is, assigning the values to component.Context.Globals instead of component.Globals. This is what assigning to component.Globals does indirectly, but in case there is an error there, this should be the solution.

I was not able to get this to work - `component.Context` was null.

> [UnityEngine.Scripting.Preserve] and link.xml

I tried these and neither appeared to make a difference.

**Comment by KurtGokhan** [comment]
Ok, found the issue thanks to the hints you gave me.

The issue was happening when Globals were changing after ReactUnity first rendered but before ReactUnity subscribed to changes in Globals. 

Should be fixed in `renderer@0.5.1`. Let me know if this also fixes your other issue.

**Comment by amagee** [comment]
Yep, confirm this is fixed in 0.5.1, thanks!
